### ISSUE TITLE
environment variables can’t get assign value in Task script part. Environment is still empty after MY_ENV=“VALUE“. The same works for ClusterTasks.

### ISSUE THREAD

**Issue by JiriHusak-lab** [tektoncd/operator#1420]
# Expected Behavior
This code:
MY_VALUE=“AAA“
export MY_VALUE
echo "MY_VALUE: $MY_VALUE"
if [ „$MY_VALUE“ == „AAA“ ]
then 
       echo „is AAA“
else
      echo „is NOT AAA“
fi

should show:
MY_VALUE: AAA
is AAA

# Actual Behavior
shows:
MY_VALUE:
is NOT AAA

# Steps to Reproduce the Problem

1.
2.
3.

# Additional Info

- Kubernetes version:
My environment:
CRC version: 2.10.2+1bc2e5a
OpenShift version: 4.11.7
Podman version: 4.2.0

Red Hat OpenShift Pipelines 1.9.2
Components
•	Tekton Pipelines: v0.41.1
•	Tekton Triggers: v0.22.1
•	Pipelines as Code: 0.15.5
•	Tekton Chains (tech-preview): v0.13.1
•	Tekton Hub (tech-preview): v1.11.3

  **Output of `kubectl version`:**

  ```
  (paste your output here)
  ```

- Tekton Pipeline version:

  **Output of `tkn version` or `kubectl get pods -n tekton-pipelines -l app=tekton-pipelines-controller -o=jsonpath='{.items[0].metadata.labels.version}'`**

I have a critical issue. Believe me I tried mostly everything. 
When Tekton pipeline executes Task, I can’t work with environment variables – I mean to set ENV variable in Task script part and use it.
Parameters work, env variables in ClusterTasks WORKS! Yes, in cluster tasks it works.
What doesn’t work:
MY_VALUE=“AAA“
export MY_VALUE
echo $MY_VALUE
if [ „$MY_VALUE“ == „AAA“ ]
then 
       echo „is AAA“
else
      echo „is NOT AAA“
fi
Even echo prints nothing, variable is empty, test shows is NOT AAA.
I’ve put sleep 600 into the script, use Terminal for the same task pod and it works there! So in the same pod, when I eneter POD via Terminal, I can do MY_VALUE=“AAA“; echo $MY_VALUE.
I tried different images for Task, for example:
registry.access.redhat.com/ubi8/ubi-minimal:8.2
registry.access.redhat.com/ubi8/ubi@sha256:e721f98a49e731f0bd64f8e89c229e1dbb38c46265d92849b3e0bedaf5f81139
mcr.microsoft.com/playwright:focal

Behaviour is always the same.
I’m really confused.
Why in cluster tasks it works??? I’ve modified cluster tasks and at the beginning of the script section I’ve put MY_VALUE=“AAA“; echo $MY_VALUE. It works.
For tasks it doesn’t work regardless image.
For the same task POD, when enetered via Terminal, it works.
Sounds crazy.
How Tekton could change behaviour of POD when run in pipeline?
Issue is critical, I can’t use some Tasks.

And for example, also this Task fails because of the same reason. It is not possible to set values into env variables in script part.

https://raw.githubusercontent.com/tektoncd/catalog/main/task/sonarqube-scanner/0.2/sonarqube-scanner.yaml

Of course, it is hardly reproducable for you, It is specific to my environment. But tell me HOW, how this strange functionality could be achieved? Just theoretically, in case this behaviour would be intentional requirement, how could Tekton influence shell run in Task container in the way, that environment variable can't get value assigned?
And it works in all Cluster Tasks!

My colleague is using Openshift v4.10.52 and openshift-pipelines-operator-rh.v1.8.2 and it is working for him corretcly.

Thanks.
Jiri

**Comment by JiriHusak-lab** [comment]
Ok, I’ve located the rootcause. Am I doing something wrong, when creating pipeline run using „oc create -f filename.yaml“ or using „oc apply -f filename.yaml“ ?

So, issue seems to be that when I create pipeline run using:
oc apply -f ci.PipelineRun.yaml
or
oc create -f ci.PipelineRun.yaml
then created pipeline run yaml have replaced } character with ( character.
This replacement, when looking into create pipeline run yaml code doesn’t happen for cluster tasks. And it explains why cluster tasks are not facing this bug.

Example of wrongly created part of pipeline run code (Task part) using this creation way:
echo "STARTING test-task-script
-------------------------------------------------------"

            echo " "

            M_VALUE="AAA"

            echo "M_VALUE: $(M_VALUE)" $(M_VALUE)

            if [ "${M_VALUE}" != "AAA" ] ; then
               echo "IS NOT AAA"
               echo "M_VALUE: $(M_VALUE)" $(M_VALUE)
            else
               echo "AAA"
               echo "IS AAA"
               echo "M_VALUE: $(M_VALUE)" $(M_VALUE)
            fi
so logs of run looks like this:
STEP-TEST-TASK-SCRIPT
STARTING test-task-script -------------------------------------------------------

/tekton/scripts/script-0-9qz65: line 6: M_VALUE: command not found
M_VALUE:
/tekton/scripts/script-0-9qz65: line 6: M_VALUE: command not found
AAA
IS AAA
/tekton/scripts/script-0-9qz65: line 13: M_VALUE: command not found
/tekton/scripts/script-0-9qz65: line 13: M_VALUE: command not found
M_VALUE:

When Pipeline started from oc console, using GUI for pipeline start,
Then
Code looks like this:
script: >
echo "STARTING test-task-script
-------------------------------------------------------"

      echo " "

      M_VALUE="AAA"

      echo "M_VALUE: ${M_VALUE}"

      if [ "${M_VALUE}" == "AAA" ] ; then
         echo "IS AAA"
         echo "M_VALUE: ${M_VALUE}" ${M_VALUE}
      else
         echo "IS NOT AAA"
         echo "M_VALUE: ${M_VALUE}" ${M_VALUE}
      fi
so , correctly and log looks correctly as well:
STEP-TEST-TASK-SCRIPT
STARTING test-task-script -------------------------------------------------------

M_VALUE: AAA
IS AAA
M_VALUE: AAA AAA

Original scripts (3 scripts):
1st script) ci.PipelineRun.yaml:
1 apiVersion: tekton.dev/v1beta1
2 kind: PipelineRun
3 metadata:
4 name: test-tekton
5 spec:
6 taskRunSpecs:
7 pipelineRef:
8 name: test-tekton

2nd script) ci.Pipeline.yaml
1 apiVersion: tekton.dev/v1beta1
2 kind: Pipeline
3 metadata:
4 name: test-tekton
5 spec:
6 tasks:
7 # =========================================== TEST-TASK-SCRIPT ================================================
8 - name: test-task-script
9 taskRef:
10 name: test-task-script

3rd script) test-tekton.Task.yaml
1 apiVersion: tekton.dev/v1beta1
2 kind: Task
3 metadata:
4 name: test-task-script
5 spec:
6 steps:
7 - name: test-task-script
8 image: registry.access.redhat.com/ubi8/ubi@sha256:e721f98a49e731f0bd64f8e89c229e1dbb38c46265d92849b3e0bedaf5f81139
9 script: |
10 echo "STARTING test-task-script -------------------------------------------------------"
11 echo " "
12 M_VALUE="AAA"
13 echo "M_VALUE: ${M_VALUE}"
14 if [ "${M_VALUE}" == "AAA" ] ; then
15 echo "IS AAA"
16 echo "M_VALUE: ${M_VALUE}" ${M_VALUE}
17 else
18 echo "IS NOT AAA"
19 echo "M_VALUE: ${M_VALUE}" ${M_VALUE}
20 fi

**Comment by JiriHusak-lab** [comment]
So now question. Regarding my previous comment explaining bug rootcause, is it issue of OpenShift, CRC or Tekton? To whom log the bug?

**Comment by vdemeester** [comment]
> So, issue seems to be that when I create pipeline run using:
> oc apply -f ci.PipelineRun.yaml
> or
> oc create -f ci.PipelineRun.yaml
> then created pipeline run yaml have replaced } character with ( character.

This sounds really weird 🤔. What `oc` version are you using ? At least this is not a Tekton issue as it works as intended (provided with the correct yaml) — if something mutate the yaml prior to being created in the cluster, Tekton cannot do anything about that.

**Comment by JiriHusak-lab** [comment]
Yes, it is. I've just noticed that the same happens when creating Task using oc apply -f or oc create -f.
${M_VALUE} is replaced with $(M_VALUE)
$M_VALUE is replaced to empty string: 
if [ "$M_VALUE" != "AAA" ] ; then 
is created as
if [ "" != "AAA" ] ; then

So points to oc CLI binary or CRC Openshift?
oc version
Client Version: 4.11.7
Kustomize Version: v4.5.4
Server Version: 4.11.7
Kubernetes Version: v1.24.0+3882f8f

**Comment by vdemeester** [comment]
@JiriHusak-lab are you using `cat <<EOF | oc create -f - ` ? Because if you do, bash is most likely going to interpret things and thus replace something like `$M_VALUE` with the actualy value.

What is the exact command you are running ?

**Comment by JiriHusak-lab** [comment]
Bingo, great point. Yes, on real project , your explanations is correct, command is:
export PIPELINE_RUN_NAME=$(envsubst < "ci_cd/tekton/runs/ci.PipelineRun.yaml" | oc create -n $K8S_NAMESPACE -f - --output=jsonpath='{.metadata.name}')
So this is what I can change.  Thanks a lot! I was really mad already. Thanks.

Now I need to elaborate why my simple sample did replacement as well when I used oc apply -f / oc create -f only.
Thanks again.
